safehttp/plugins/cors/cors_test.go

```diff
@@ -0,0 +1,574 @@
+			}(),
+			want: map[string][]string{
+				"Access-Control-Allow-Origin": {"https://foo.com"},
+				"Vary":                        {"Origin"},
+			},
+		},
+		{
+			name: "No Origin header",
+			req: func() *safehttp.IncomingRequest {
+				r := safehttptest.NewRequest(safehttp.MethodPut, "http://bar.com", nil)
+				r.Header.Set("X-Cors", "1")
+				r.Header.Set("Content-Type", "application/json")
+				return r
+			}(),
+			want: map[string][]string{},
+		},
+		{
+			name: "AllowCredentials but no cookies",
+			req: func() *safehttp.IncomingRequest {
+				r := safehttptest.NewRequest(safehttp.MethodPut, "http://bar.com", nil)
+				r.Header.Set("Origin", "https://foo.com")
+				r.Header.Set("X-Cors", "1")
+				r.Header.Set("Content-Type", "application/json")
+				return r
+			}(),
+			allowCredentials: true,
+			want: map[string][]string{
+				"Access-Control-Allow-Origin": {"https://foo.com"},
+				"Vary":                        {"Origin"},
+			},
+		},
+		{
+			name: "AllowCredentials with cookies",
+			req: func() *safehttp.IncomingRequest {
+				r := safehttptest.NewRequest(safehttp.MethodPut, "http://bar.com", nil)
+				r.Header.Set("Origin", "https://foo.com")
+				r.Header.Set("X-Cors", "1")
+				r.Header.Set("Content-Type", "application/json")
+				r.Header.Set("Cookie", "a=b")
+				return r
+			}(),
+			allowCredentials: true,
+			want: map[string][]string{
+				"Access-Control-Allow-Credentials": {"true"},
+				"Access-Control-Allow-Origin":      {"https://foo.com"},
+				"Vary":                             {"Origin"},
+			},
+		},
+		{
+			name: "Expose one header",
+			req: func() *safehttp.IncomingRequest {
+				r := safehttptest.NewRequest(safehttp.MethodPut, "http://bar.com", nil)
+				r.Header.Set("Origin", "https://foo.com")
+				r.Header.Set("X-Cors", "1")
+				r.Header.Set("Content-Type", "application/json")
+				return r
+			}(),
+			exposedHeaders: []string{"Aaaa"},
+			want: map[string][]string{
+				"Access-Control-Expose-Headers": {"Aaaa"},
+				"Access-Control-Allow-Origin":   {"https://foo.com"},
+				"Vary":                          {"Origin"},
+			},
+		},
+		{
+			name: "Expose multiple headers",
+			req: func() *safehttp.IncomingRequest {
+				r := safehttptest.NewRequest(safehttp.MethodPut, "http://bar.com", nil)
+				r.Header.Set("Origin", "https://foo.com")
+				r.Header.Set("X-Cors", "1")
+				r.Header.Set("Content-Type", "application/json")
+				return r
+			}(),
+			exposedHeaders: []string{"Aaaa", "Bbbb", "Cccc"},
+			want: map[string][]string{
+				"Access-Control-Expose-Headers": {"Aaaa, Bbbb, Cccc"},
+				"Access-Control-Allow-Origin":   {"https://foo.com"},
+				"Vary":                          {"Origin"},
+			},
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			rr := safehttptest.NewResponseRecorder()
+
+			it := cors.Default("https://foo.com")
+			it.AllowCredentials = tt.allowCredentials
+			it.ExposedHeaders = tt.exposedHeaders
+			it.Before(rr.ResponseWriter, tt.req)
+
```

Comment:
Why is the status code not checked?

Reply:
Well this path through the plugin never sets a status, it just returns a `safehttp.Result{}`, leaving the setting of the status up to later plugins or the handler. So there isn't really anything to check. We should probably check that the status is unset, but currently the `safehttptest.ResponseRecorder` sets 200 OK as the default response status. Maybe we should change this to be 0 instead? So that we could check for these cases.

Reply:
Hmm, I think that'd be a good idea to also avoid future bugs (e.g. we expect something to set the status code to 200 OK, but in fact it doesn't and tests pass because  `safehttptest.ResponseRecorder` AND `safehttp_test.responseRecorder` set the status to 200 OK by default). However, I think that should be in a separate clean-up PR not here.

Reply:
Opened #152